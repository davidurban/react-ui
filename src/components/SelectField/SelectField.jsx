import PropTypes from 'prop-types';
import React from 'react';
import styles from './SelectField.scss';

const SelectField = (props) => {
  const {
    changeHandler,
    description,
    error,
    fieldId,
    isLabelVisible,
    options,
    ...otherProps
  } = props;

  let labelClass = styles.label;
  if (isLabelVisible) {
    if (props.required) {
      labelClass = styles.isLabelRequired;
    }
  } else {
    labelClass = styles.isLabelHidden;
  }

  let rootClass = styles.root;
  if (isLabelVisible) {
    if (props.disabled) {
      rootClass = styles.isRootDisabled;
    }
  } else {
    rootClass = styles.isRootCondensed;
  }

  return (
    <div className={rootClass}>
      <label htmlFor={fieldId}>
        <div className={labelClass}>
          {props.label}
        </div>
        <select
          id={fieldId}
          {...otherProps}
          className={props.error ? styles.isSelectInvalid : styles.select}
          onChange={changeHandler}
        >
          {
            options.map(option => (
              <option
                key={option.value}
                value={option.value}
              >
                {option.label}
              </option>
            ))
          }
        </select>
      </label>
      {description && (
        <div className={styles.description}>
          {description}
        </div>
      )}
      {error && (
        <div className={styles.error}>
          {error}
        </div>
      )}
    </div>
  );
};

SelectField.defaultProps = {
  changeHandler: null,
  description: null,
  disabled: false,
  error: null,
  isLabelVisible: true,
  required: false,
  value: undefined,
};

SelectField.propTypes = {
  changeHandler: PropTypes.func,
  description: PropTypes.string,
  disabled: PropTypes.bool,
  error: PropTypes.string,
  fieldId: PropTypes.string.isRequired,
  isLabelVisible: PropTypes.bool,
  label: PropTypes.string.isRequired,
  options: PropTypes.arrayOf(PropTypes.shape({
    label: PropTypes.string.isRequired,
    value: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
    ]),
  })).isRequired,
  required: PropTypes.bool,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
};

export default SelectField;
