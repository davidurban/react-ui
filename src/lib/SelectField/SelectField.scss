@import '../Theme/icons';
@import '../Theme/mixins';
@import '../Theme/variables';
@import '../Theme/root';

.root {
  position: relative;
  display: flex;
  flex-direction: column;
  box-sizing: border-box;
  padding: 0 0.25rem 0 0;
  margin-bottom: 1rem;

  &::after {
    content: '';
    position: absolute;
    top: 2rem + ($rui-input-height - ($rui-dropdown-caret-icon-dimension / 2)) / 2;
    right: 0.5rem;
    width: $rui-dropdown-caret-icon-dimension;
    height: $rui-dropdown-caret-icon-dimension;
    margin: 0;
    vertical-align: middle;
    background-image: generatecareticon($rui-select-after-color);
    background-size: contain;
    background-repeat: no-repeat;
  }

  &:hover::after {
    color: $rui-select-after-hover-color;
  }
}

.isRootDisabled {
  @include rui-disabled-form-fields();

  &:hover::after {
    color: $rui-select-after-color;
  }
}

.isRootCondensed {
  composes: root;

  &::after {
    top: 0.25rem + ($rui-input-height - ($rui-dropdown-caret-icon-dimension / 2)) / 2;
  }
}

.label {
  display: inline-block;
  margin-bottom: 0.25rem;
}

.isLabelRequired {
  composes: label;

  &::after {
    content: ' *';
  }
}

.isLabelHidden {
  position: absolute;
  width: 1px;
  height: 1px;
  padding: 0;
  overflow: hidden;
  clip: rect(0, 0, 0, 0);
  white-space: nowrap;
  border: 0;
}

.select {
  @include rui-custom-inputs($rui-input-height);
  @include rui-form-fields-appearance(2.5rem, $rui-form-field-border-color);
  @include rui-input-border-bottom-animation($rui-input-border-color);

  position: relative;
  z-index: 2;
  box-sizing: border-box;
  width: 100%;
  padding: 0 ($rui-dropdown-caret-icon-dimension + 0.5rem) 0 0.25rem;
  font-family: var(--font-family);
  cursor: pointer;

  &:hover,
  &:active {
    border-color: $rui-select-hover-background;
    background: $rui-select-hover-background;
  }

  &:disabled {
    @include rui-disabled-form-fields();

    &:hover {
      border-color: $rui-form-field-border-color;
      background: none;
    }
  }
}

.select::-ms-expand {
  display: none;
}

.select > option {
  color: $rui-select-option-font-color;
}

.isSelectInvalid {
  composes: select;
  border-bottom: $rui-border-width solid $rui-input-invalid-color;
  background: linear-gradient(to bottom, transparent 95%, $rui-input-invalid-color 95%);
}

.description,
.error {
  font-size: 0.8rem;
}

.error {
  min-height: 1rem;
  color: $rui-input-invalid-color;
}
