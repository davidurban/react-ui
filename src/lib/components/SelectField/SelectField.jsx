import PropTypes from 'prop-types';
import React from 'react';
import styles from './SelectField.scss';

const SelectField = (props) => {
  let labelClass = styles.label;
  if (props.isLabelVisible) {
    if (props.required) {
      labelClass = styles.isLabelRequired;
    }
  } else {
    labelClass = styles.isLabelHidden;
  }

  let rootClass = styles.root;
  if (props.isLabelVisible) {
    if (props.disabled) {
      rootClass = styles.isRootDisabled;
    }
  } else {
    rootClass = styles.isRootCondensed;
  }

  return (
    <div className={rootClass}>
      <label htmlFor={props.fieldId}>
        <div className={labelClass}>
          {props.label}
        </div>
        <select
          id={props.fieldId}
          disabled={props.disabled}
          required={props.required}
          value={props.value}
          className={props.error ? styles.isSelectInvalid : styles.select}
          onChange={props.changeHandler}
        >
          {
            props.options.map(option => (
              <option
                key={option.value}
                value={option.value}
              >
                {option.label}
              </option>
            ))
          }
        </select>
      </label>
      {props.description && (
        <div className={styles.description}>
          {props.description}
        </div>
      )}
      {props.error && (
        <div className={styles.error}>
          {props.error}
        </div>
      )}
    </div>
  );
};

SelectField.defaultProps = {
  changeHandler: null,
  description: null,
  disabled: false,
  error: null,
  isLabelVisible: true,
  required: false,
  value: undefined,
};

SelectField.propTypes = {
  changeHandler: PropTypes.func,
  description: PropTypes.string,
  disabled: PropTypes.bool,
  error: PropTypes.string,
  fieldId: PropTypes.string.isRequired,
  isLabelVisible: PropTypes.bool,
  label: PropTypes.string.isRequired,
  options: PropTypes.arrayOf(PropTypes.shape({
    label: PropTypes.string.isRequired,
    value: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
    ]),
  })).isRequired,
  required: PropTypes.bool,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
};

export default SelectField;
