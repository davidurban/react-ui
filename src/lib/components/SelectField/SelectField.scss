@import '../../shared/settings/borders';
@import '../../shared/settings/forms.deprecated';
@import '../../shared/settings/typography';
@import '../../shared/tools/forms';
@import '../../shared/tools/icons';

.root {
  position: relative;
  display: flex;
  flex-direction: column;
  box-sizing: border-box;
  padding: 0 0.25rem 0 0;
  margin-bottom: 1rem;

  &::after {
    content: '';
    position: absolute;
    top: 2rem + ($form-field-height - ($select-dropdown-caret-icon-size / 2)) / 2;
    right: 0.5rem;
    width: $select-dropdown-caret-icon-size;
    height: $select-dropdown-caret-icon-size;
    margin: 0;
    vertical-align: middle;
    background-image: icon-generate-caret($select-after-color);
    background-size: contain;
    background-repeat: no-repeat;
  }

  &:hover::after {
    color: $select-after-hover-color;
  }
}

.isRootDisabled {
  @include form-disabled-state();

  &:hover::after {
    color: $select-after-color;
  }
}

.isRootCondensed {
  composes: root;

  &::after {
    top: 0.25rem + ($form-field-height - ($select-dropdown-caret-icon-size / 2)) / 2;
  }
}

.label {
  display: inline-block;
  margin-bottom: 0.25rem;
}

.isLabelRequired {
  composes: label;

  &::after {
    content: ' *';
  }
}

.isLabelHidden {
  position: absolute;
  width: 1px;
  height: 1px;
  padding: 0;
  overflow: hidden;
  clip: rect(0, 0, 0, 0);
  white-space: nowrap;
  border: 0;
}

.select {
  @include form-custom-inputs($form-field-height);
  @include form-field-base(2.5rem, $form-field-border-color);
  @include form-border-bottom-animation($form-field-border-color);

  position: relative;
  z-index: 2;
  box-sizing: border-box;
  width: 100%;
  padding: 0 ($select-dropdown-caret-icon-size + 0.5rem) 0 0.25rem;
  font-family: $typography-font-family-base;
  cursor: pointer;

  &::-ms-expand {
    display: none;
  }

  &:hover,
  &:active {
    border-color: $select-hover-background;
    background: $select-hover-background;
  }

  &:disabled {
    @include form-disabled-state();

    &:hover {
      border-color: $form-field-border-color;
      background: none;
    }
  }
}

.select > option {
  color: $select-option-text-color;
}

.isSelectInvalid {
  composes: select;
  border-bottom: $border-width solid $form-invalid-color;
  background: linear-gradient(to bottom, transparent 95%, $form-invalid-color 95%);
}

.description,
.error {
  font-size: 0.8rem;
}

.error {
  min-height: 1rem;
  color: $form-invalid-color;
}
