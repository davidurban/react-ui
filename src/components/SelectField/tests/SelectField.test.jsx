import React from 'react';
import { shallow, mount } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import sinon from 'sinon';
import SelectField from '../index';

describe('rendering', () => {
  it('renders correctly mandatory props only', () => {
    const tree = shallow(<SelectField
      fieldId="test"
      label="label"
      options={[
        {
          label: 'choice 1',
          value: 'ch1',
        },
        {
          label: 'choice 2',
          value: 'ch2',
        },
      ]}
    />);

    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly only one option', () => {
    const tree = shallow(<SelectField
      fieldId="test"
      label="label"
      options={[
        {
          label: 'choice 1',
          value: 'ch1',
        },
      ]}
    />);

    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly with all props', () => {
    const tree = shallow(<SelectField
      label="label"
      isLabelVisible={false}
      disabled
      fieldId="test"
      value="ch1"
      description="some help"
      error="some error"
      required
      options={[
        {
          label: 'choice 1',
          value: 'ch1',
        },
        {
          label: 'choice 2',
          value: 'ch2',
        },
      ]}
    />);

    expect(shallowToJson(tree)).toMatchSnapshot();
  });
});

describe('functionality', () => {
  it('calls changeHandler() on changing selected option', () => {
    const spy = sinon.spy();
    const component = mount(<SelectField
      label="label"
      fieldId="id"
      options={[
        {
          label: 'choice 1',
          value: 'ch1',
        },
        {
          label: 'choice 2',
          value: 'ch2',
        },
      ]}
      changeHandler={spy}
      value="ch1"
    />);

    component
      .find('select')
      .simulate('change', {
        preventDefault: () => {},
        target: { value: 'ch2' },
      });
    expect(spy.calledOnce).toEqual(true);
  });
});
