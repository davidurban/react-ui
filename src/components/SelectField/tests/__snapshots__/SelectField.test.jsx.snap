// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly mandatory props only 1`] = `
<div
  className="root"
>
  <label
    htmlFor="test"
  >
    <div
      className="label"
    >
      label
    </div>
    <select
      className="select"
      disabled={false}
      id="test"
      label="label"
      onChange={null}
      required={false}
    >
      <option
        key="ch1"
        value="ch1"
      >
        choice 1
      </option>
      <option
        key="ch2"
        value="ch2"
      >
        choice 2
      </option>
    </select>
  </label>
</div>
`;

exports[`rendering renders correctly only one option 1`] = `
<div
  className="root"
>
  <label
    htmlFor="test"
  >
    <div
      className="label"
    >
      label
    </div>
    <select
      className="select"
      disabled={false}
      id="test"
      label="label"
      onChange={null}
      required={false}
    >
      <option
        key="ch1"
        value="ch1"
      >
        choice 1
      </option>
    </select>
  </label>
</div>
`;

exports[`rendering renders correctly with all props 1`] = `
<div
  className="isRootCondensed"
>
  <label
    htmlFor="test"
  >
    <div
      className="isLabelHidden"
    >
      label
    </div>
    <select
      className="isSelectInvalid"
      disabled={true}
      id="test"
      label="label"
      onChange={null}
      required={true}
      value="ch1"
    >
      <option
        key="ch1"
        value="ch1"
      >
        choice 1
      </option>
      <option
        key="ch2"
        value="ch2"
      >
        choice 2
      </option>
    </select>
  </label>
  <div
    className="description"
  >
    some help
  </div>
  <div
    className="error"
  >
    some error
  </div>
</div>
`;
